wavelength-plugins/friends.js

```diff
@@ -0,0 +1,148 @@
+/*
+ * FC Friends
+ * By WGC
+*/
+'use strict';
+
+function genFC() {
+	let fc = Math.floor(Math.random() * 89999999 + 10000000);
+	let FCjson = Db.fc.keys();
+	let finalFC = 'WL' + fc;
+	for (let u of FCjson) {
+		while (FCjson[u] === [finalFC]) {
+			fc = Math.floor(Math.random() * 89999999 + 10000000);
+			finalFC = 'WL' + fc;
+		}
+	}
+	return finalFC;
+}
+
+const FC = {
+	getFC: function (user) {
+		user = toId(user);
+		let fcget = Db.fc.get(user);
+		return fcget;
+	},
+
+	addFC: function (user, fc) {
+		user = toId(user);
+		Db.fc.set(user, fc);
+		return this.sendReply('Your friend code has been added to the system. Give to others to add you as a friend!');
+	},
+};
+
+const Friend = {
+	addFriend: function (target, fc, user) {
+		target = toId(target);
+		fc = fc.toLowerCase();
+		let thisFC = FC.getFC(target);
+		thisFC = fc.toLowerCase();
+		if (fc === thisFC) {
+			if (Db.friend.get(user, []).length === 0) Db.friend.set(user, []);
+			Db.friend.set(Db.friend.get(user).push(target));
+			user.popup(`You have added ${target} as a friend!`);
+			return;
+		} else {
+			user.popup(`${fc} is not the right friendcode for ${target}.`);
+			return;
+		}
+	},
+
+	removeFriend: function (target, user) {
+		target = toId(target);
+		if (!Db.friend.get(user).includes(target)) return this.errorReply('You are not friends with this user.');
+		let array = Db.friend.get(user);
+		let tot = 0;
+		for (let i = 0; i < array.length; i++) {
+			if (array[i] !== target) {
+				tot = tot + 1;
+			} else {
+				i = array.length;
+			}
+		}
+		Db.friends.set(Db.friends.get(user).splice(tot, 1));
+		return this.sendReply(`You are no longer friends with ${target}.`);
```

Comment:
Once again, `Friends` does not have the same `this` context as a command does; therefore, we cannot access `this.errorReply` or any prototype methods of `Chat.CommandContext`.

Whereas we used `Array.prototype.some` to check an array to see if any of the array's elements match `true` against the provided function, we can use `Array.prototype.findIndex` to find the index in which `target` is in the array.

Par exemple,
```js
let userid = user.userid; // helpful alias
let curFriends = Db.friend.get(userid, []);
if (!curFriends.length) return user.popup("You don't have any friends to remove.");
let targetIndex = curFriends.findIndex(curFriend => toId(target) === curFriend);
if (targetIndex <= 0) return user.popup("You are not friends with this user.");
curFriends.splice(targetIndex, 1);
Db.friend.set(userid, curFriends);
user.popup(`You are no longer friends with "${target}".`);